### DISCUSSION TITLE
How to link to correctionlib from a CMSSW module

### DISCUSSION THREAD

**Discussion by nsmith-** [cms-nanoAOD/correctionlib#86] [Q&A] [answered]
Inside CMSSW, some may want to use an EDAnalyzer or other framework plugin with correctionlib.

**Comment by nsmith-** [comment, marked as answer]
In the long term, the package will be part of cmsdist, and in fact v1.1 [is already in](https://github.com/cms-sw/cmsdist/pull/6659) the gcc900 externals: `/cvmfs/cms.cern.ch/slc7_amd64_gcc900/external/py3-correctionlib`, as used in CMSSW 11_3_X releases. In that case, it should suffice to add the line
```xml
<use name="correctionlib"/>
```
to the package `BuildFile.xml` to use it.

If one wants a newer version than what is available in the externals, the following recipe may work:
1. First check if `scram tool info correctionlib` is an acceptable version, if so skip to the last step.
2. Install correctionlib, preferably in a way that is portable with your CMSSW environment (e.g. `pip install --prefix $CMSSW_BASE/local` or use [scram-pip](https://github.com/FNALLPC/lpc-scripts/blob/master/scram-pip))
3. Create an appropriate scram tool description, e.g. using this bash script
```bash
#!/bin/bash

cat <<EOF > correctionlib.xml
<tool name="correctionlib" version="$(correction config --version)">
  <lib name="correctionlib"/>
  <client>
    <environment name="INCLUDE" default="$(correction config --incdir)"/>
    <environment name="LIBDIR" default="$(correction config --libdir)"/>
  </client>
</tool>
EOF
```
4. Instal the new tool with `scram setup correctionlib.xml` (`scram tool remove correctionlib` to remove)
5. Add the line
```xml
<use name="correctionlib"/>
```
to your package `BuildFile.xml`

**Reply by nsmith-** [reply]
A confounding issue is that there is ABI incompatibility issues due to `_GLIBCXX_USE_CXX11_ABI` being set to 1 in CMSSW / any modern gcc version, and the fact that python manylinux wheels have it set to 0 (since they use RHEL7 stock gcc). The workaround is to make sure that correctionlib is compiled using the gcc environment of CMSSW, e.g. after sourcing a CMSSW release make sure to pip install from source.

At present, the source distribution on pypi seems to be broken, but installation from git seems to work OK, e.g.
`python3 -m pip install --user git+https://github.com/cms-nanoAOD/correctionlib.git@2.0.0rc6`

**Reply by nsmith-** [reply]
Small update: the pypi source distribution is now functional, and:
```
python3 -m pip install --user --no-binary=correctionlib correctionlib
```
should work to install correctionlib from source.

**Reply by makortel** [reply]
I _think_ the recipe needs `cmsenv` after `scram setup correctionlib.xml`.

**Comment by nsmith-** [comment]
Another option is to use pip-in-scram, as developed in https://github.com/FNALLPC/lpc-scripts/blob/master/scram-pip, also available at `/cvmfs/cms-lpc.opensciencegrid.org/FNALLPC/lpc-scripts/scram-pip`

**Comment by pieterdavid** [comment]
In case it helps anyone: I turned the above instructions into a [script](https://gist.github.com/pieterdavid/8f43f302e9f8a71f92702101600b7ddb) (tested with CMSSW_10_6_29 for both python2 and python3)

**Reply by nsmith-** [reply]
Nice! Just a comment, some of the vars in the script could be determined from `correction config` (in fact we could add a whole buildfile maker into that?)

**Reply by pieterdavid** [reply]
yes, but not for python2, as far as I understood - in any case this is a temporary solution for things like nanoAODtools development or comparisons in 10_6
